### PR TITLE
3DS Result Object Refactor

### PR DESCRIPTION
**Pull request by sarahkoop**
### Summary of changes

 - Refactor all public methods in 3DS module to have single parameter callbacks with single result objects.
 - This is a bit more complex than other modules because the previous `ThreeDSecureResult` object was used throughout the flow to contain the lookup and nonce after the initial lookup call, and after the completion of authentication flow. 
 - Update to make the previous `ThreeDSecureResult` object into a `ThreeDSecureParams` class - used to launch the `ThreeDSecureLauncher`, and made all fields except `ThreeDSecureLookup` private on that class (so merchants can still access intermediate lookup params if desired). 
 - Added `ThreeDSecurePaymentAuthRequest` object with a `ReadyToLaunch` state that contains the `ThreeDSecureParams`, and a `LaunchNotRequired` state that represents a successful 3DS flow where no additional authentication challenge was required. The `LaunchNotRequired` contains `nonce` and `lookup`. The `nonce` and `lookup` also do exist on the `ThreeDSecureParams` object, but for integration clarity they are separated. 
 - Add `ThreeDSecureResult` for the completion of tokenization (full auth challenge completion), which just returns a `nonce` on success - through testing it appears that we don't receive the `lookup` from this API call, so it is not returned again here.

Open to suggestions on any changes to make the integration pattern more clear

 ### Checklist

 - [x] Added a changelog entry

### Authors

- @sarahkoop

### CODE REVIEW

File: ThreeDSecure/src/main/java/com/braintreepayments/api/ThreeDSecureClient.java [hunk 1]
```diff
@@ -282,22 +281,23 @@ public void tokenize(ThreeDSecurePaymentAuthResult paymentAuthResult,
...
-                    api.authenticateCardinalJWT(threeDSecureResult, jwt,
+                    api.authenticateCardinalJWT(threeDSecureParams, jwt,
                             (threeDSecureResult1, error) -> {
                                 if (threeDSecureResult1 != null) {
                                     if (threeDSecureResult1.hasError()) {
                                         braintreeClient.sendAnalyticsEvent(
                                                 "three-d-secure.verification-flow.upgrade-payment-method.failure.returned-lookup-nonce");
-                                    } else {
+                                        callback.onThreeDSecureResult(new ThreeDSecureResult.Failure(new BraintreeException(threeDSecureResult1.getErrorMessage()), threeDSecureResult1.getThreeDSecureNonce()));
+                                    } else if (threeDSecureResult1.getThreeDSecureNonce() != null){
```
Comment (sshropshire) [inline ThreeDSecure/src/main/java/com/braintreepayments/api/ThreeDSecureClient.java:291]:
```suggestion
                                    } else if (threeDSecureResult1.getThreeDSecureNonce() != null) {
```

File: BraintreeCore/schemas/com.braintreepayments.api.AnalyticsDatabase/1.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{
+  "formatVersion": 1,
```
[the remark this answers is not in the document]
Comment (sarahkoop) [reply BraintreeCore/schemas/com.braintreepayments.api.AnalyticsDatabase/1.json]: Ah no - it was merged in with another PR so now removed 👍

**Review by sshropshire — APPROVED** [review]
🌊

File: v5_MIGRATION_GUIDE.md [hunk 3]
```diff
@@ -202,9 +202,11 @@ class MyActivity : FragmentActivity() {
...
 -       initializeClients()
 +       threeDSecureLauncher = ThreeDSecureLauncher(this) { paymentAuthResult ->
-+            threeDSecureClient.tokenize(paymentAuthResult) { threeDSecureResult, error ->
-+                error?.let { /* handle error */ }
-+                threeDSecureResult?.let { /* handle threeDSecureResult.tokenizedCard */ }
++            threeDSecureClient.tokenize(paymentAuthResult) { result ->
++               when (result) {
++                   is ThreeDSecureResult.Success -> { /* send result.nonce to server */}
++                   is ThreeDSecureResult.Failure -> { /* handle result.error */}
++                   is ThreeDSecureResult.Cancle -> { /* user canceled authentication */}
```
Comment (tdchow) [inline v5_MIGRATION_GUIDE.md:209]:
```suggestion
+                   is ThreeDSecureResult.Cancel -> { /* user canceled authentication */}
```

**Review by tdchow — COMMENTED** [review]
Loving the single result object in favor of the result and error params 🙌

**Review by tdchow — APPROVED** [review]

**Pull request merged**